### PR TITLE
add kafka adapter

### PR DESCRIPTION
**Pull request by akashg3627**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -19,6 +19,7 @@ pydantic = ">=1.10.12,<1.11.0"
 nats-py = { version = ">=2.3.1,<2.4.0", optional = true }
 aiohttp = { version = ">=3.8.5,<3.9.0", optional = true }
 boto3 =  { version = ">=1.28.16,<1.29.0", optional = true }
+kafka-python = "^2.0.2"
```
Comment (chiragjn) [inline pyproject.toml:22]: This should be optional like nats and sqs groups
See lines 19-26

Comment (akashg3627) [reply pyproject.toml:22]: sure, will do that

File: async_processor/kafka_pub_sub.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            try:
+                res = await run_in_threadpool(
+                    self._consumer.poll, timeout_ms=1000, max_records=1
+                )
+            except Exception as ex:
+                raise InputMessageFetchFailure() from ex
+            if len(res.keys()) == 0:
+                logger.debug("No messages in queue")
+                continue
+            for _, msgs in res.items():
```
Comment (debajyoti-truefoundry) [inline async_processor/kafka_pub_sub.py:52]: We need to ack the message manually.

File: async_processor/kafka_pub_sub.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    Output,
+)
+
+
+class KafkaInput(Input):
+    def __init__(self, config: KafkaInputConfig):
+        self._bootstrap_servers = config.bootstrap_servers
+        self._topic_name = config.topic_name
+        self._auth = config.auth
+        self._consumer = KafkaConsumer(
```
Comment (debajyoti-truefoundry) [inline async_processor/kafka_pub_sub.py:22]: We need to understand if we need to pass group id explicitly here.

Comment (debajyoti-truefoundry) [reply async_processor/kafka_pub_sub.py:22]: And what happens if we do not pass it.

File: pyproject.toml [hunk 4]
```diff
@@ -19,10 +19,12 @@ pydantic = ">=1.10.12,<1.11.0"
 nats-py = { version = ">=2.3.1,<2.4.0", optional = true }
 aiohttp = { version = ">=3.8.5,<3.9.0", optional = true }
 boto3 =  { version = ">=1.28.16,<1.29.0", optional = true }
+kafka-python = { version = ">=2.0.0,<2.0.2", optional = true }
```
Comment (debajyoti-truefoundry) [inline pyproject.toml:22]:
```suggestion
kafka-python = { version = ">=2.0.0,<2.1.0", optional = true }
```

File: async_processor/kafka_pub_sub.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            sasl_plain_password=config.auth.password,
+            security_protocol="SASL_SSL",
+            sasl_mechanism="PLAIN",
+        )
+
+    @asynccontextmanager
+    async def get_input_message(
+        self,
+    ) -> AsyncIterator[Optional[str]]:
+        while True:
```
Comment (debajyoti-truefoundry) [inline async_processor/kafka_pub_sub.py:46]: Please take a look at https://github.com/truefoundry/async_service/pull/11/files#diff-ed303578d39400a4b962213796a032dfa943554525a6e15cddb4dd1e6c8a766d to see the changes we have made in the structure of the `get_input_message` method implementation.

File: async_processor/kafka_pub_sub.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    ) -> AsyncIterator[Optional[str]]:
+        try:
+            res = await run_in_threadpool(
+                self._consumer.poll,
+                timeout_ms=self.wait_time_seconds * 1000,
+                max_records=1,
+            )
+        except Exception as ex:
+            raise InputMessageFetchFailure() from ex
+        if len(res.keys()) > 0:
```
Comment (debajyoti-truefoundry) [inline async_processor/kafka_pub_sub.py:55]: We do not need this condition

File: async_processor/kafka_pub_sub.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            for _, msgs in res.items():
+                if len(msgs) > 0:
```
Comment (debajyoti-truefoundry) [inline async_processor/kafka_pub_sub.py:57]: We do not need this condition

File: async_processor/types.py [hunk 8]
```diff
@@ -100,6 +105,21 @@ def to_input(self) -> Input:
         return NATSInput(self)
 
 
+class KafkaInputConfig(InputConfig):
+    type: constr(regex=r"^kafka$") = "kafka"
+
+    bootstrap_servers: str
+    topic_name: str
+    consumer_group_id: str
+    auth: KafkaSASLAuth
```
Comment (debajyoti-truefoundry) [inline async_processor/types.py:114]: Shouldn't this be optional?

File: async_processor/types.py [hunk 9]
```diff
@@ -100,6 +105,21 @@ def to_input(self) -> Input:
         return NATSInput(self)
 
 
+class KafkaInputConfig(InputConfig):
+    type: constr(regex=r"^kafka$") = "kafka"
+
+    bootstrap_servers: str
```
Comment (debajyoti-truefoundry) [inline async_processor/types.py:111]:
```suggestion
    bootstrap_servers: Union[str, List[str]]
```

File: async_processor/types.py [hunk 10]
```diff
@@ -147,6 +167,19 @@ def to_output(self) -> Output:
         return NATSOutput(self)
 
 
+class KafkaOutputConfig(OutputConfig):
+    type: constr(regex=r"^kafka$") = "kafka"
+
+    bootstrap_servers: str
```
Comment (debajyoti-truefoundry) [inline async_processor/types.py:173]: https://github.com/truefoundry/async_service/pull/10/files#r1334075788

File: async_processor/types.py [hunk 11]
```diff
@@ -147,6 +167,19 @@ def to_output(self) -> Output:
...
+    topic_name: str
```
Comment (debajyoti-truefoundry) [inline async_processor/types.py:174]: https://github.com/truefoundry/async_service/pull/10/files#r1334075323

File: async_processor/types.py [hunk 12]
```diff
@@ -147,6 +167,19 @@ def to_output(self) -> Output:
...
+    auth: KafkaSASLAuth
```
Comment (debajyoti-truefoundry) [inline async_processor/types.py:175]: https://github.com/truefoundry/async_service/pull/10/files#r1334075323

File: async_processor/kafka_pub_sub.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        try:
+            consumer_map = await run_in_threadpool(
+                self._consumer.poll,
+                timeout_ms=self.wait_time_seconds * 1000,
+                max_records=1,
+            )
+        except Exception as ex:
+            raise InputMessageFetchFailure() from ex
+
+        if len(consumer_map) == 0:
```
Comment (debajyoti-truefoundry) [inline async_processor/kafka_pub_sub.py:68]: Is it possible that this dictionary is non-empty,

```
        for _, msgs in consumer_map.items():
            for msg in msgs:
```
But `msgs` is empty.

Comment (akashg3627) [reply async_processor/kafka_pub_sub.py:68]: no, i tested that, so if msgs is empty in that case map is also empty

File: async_processor/types.py [hunk 14]
```diff
@@ -147,6 +167,19 @@ def to_output(self) -> Output:
...
+    auth: Optional[KafkaSASLAuth] = None
+
+    def to_output(self) -> Output:
+        from async_processor.kafka_pub_sub import KafkaOutput
+
+        return KafkaOutput(self)
+
+
 class WorkerConfig(BaseModel):
     input_config: Union[
```
Comment (debajyoti-truefoundry) [inline async_processor/types.py:184]: You need add the input and output config here.

**Review by debajyoti-truefoundry — APPROVED** [review]

**Pull request merged**